File: CPP/Offer/18.cpp
struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL)
    {
    }
};

// 队列
class Solution
{
  public:
    void Mirror(TreeNode *pRoot)
    {
        if (pRoot == nullptr)
            return;
        queue<TreeNode *> qu;
        TreeNode *p = nullptr;
        qu.push(pRoot);
        while (qu.size() > 0)
        {
            p = qu.front();
            qu.pop();
            swap(p->left, p->right);
            if (p->left)
                qu.push(p->left);
            if (p->right)
                qu.push(p->right);
        }
    }
};

// 递归
class Solution
{
  public:
    void Mirror(TreeNode *pRoot)
    {
        if (!pRoot)
            return;

        // 交换左右子树
        TreeNode *pTmp;
        pTmp = pRoot->left;
        pRoot->left = pRoot->right;
        pRoot->right = pTmp;
        Mirror(pRoot->left);
        Mirror(pRoot->right);
    }
};

File: Offer/18.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# 递归
class Solution:
    # 返回镜像树的根节点
    def Mirror(self, root):
        if not root or (not root.left and not root.right):
            return root
        root.left, root.right = root.right, root.left
        self.Mirror(root.left)
        self.Mirror(root.right)

# 队列
class Solution:
    # 返回镜像树的根节点
    def Mirror(self, root):
        if not root or (not root.left and not root.right):
            return root
        queue = [root]
        while len(queue) > 0:
            p = queue.pop()
            p.left, p.right = p.right, p.left
            if p.left:
                queue.insert(0, p.left)
            if p.right:
                queue.insert(0, p.right)
